[AutoDiff] Handle `struct_element_addr` in differentiation.

dan-zheng opened the pull request:
Handle `struct_element_addr` in PrimalGen and AdjointGen.
- `struct_element_addr` makes use of `StructExtractDifferentiationStrategy`,
  along with `struct_extract`. The adjoint accumulation rules are similar.
- This enables differentiation of structs with generic members. Add tests.
- Gardening, edit doc comments.

Resolves [TF-203](https://bugs.swift.org/browse/TF-203).

dan-zheng committed:
[AutoDiff] Handle `struct_element_addr` in differentiation.

Handle `struct_element_addr` in PrimalGen and AdjointGen.
- `struct_element_addr` makes use of `StructExtractDifferentiationStrategy`,
  along with `struct_extract`. The adjoint accumulation rules are similar.
- This enables differentiation of structs with generic members. Add tests.
- Gardening, edit doc comments.

Resolves TF-203.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -573,8 +573,9 @@ class DifferentiationTask {
-  DenseMap<StructExtractInst *, StructExtractDifferentiationStrategy>
+  /// Mapping from original `struct_extract` and `struct_element_addr`
+  /// instructions to their strategies.
+  DenseMap<SILInstruction *, StructExtractDifferentiationStrategy>
```
I didn't change the names of `StructExtractDifferentiationStrategy` or `getStructExtractDifferentiationStrategies` because I couldn't think of anything more suitable.

dan-zheng committed:
Minor edits/gardening.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4092,6 +4180,93 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+                         pullbackResultBuf);
+      builder.createEndAccess(loc, readAccess, /*aborted*/ false);
+      // Deallocate pullback indirect result buffer.
+      builder.createDeallocStack(loc, pullbackResultBuf);
```
```suggestion
      builder.createDeallocStack(loc, pullbackResultBuf);
```
Please release any pullback call's indirect results, since they are immediately accumulated into another buffer.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22669/commits/25f16291c2e794966aca760347523b4862acd80b, I believe, by calling `emitCleanup` on `pullbackResultBuf`.
There's an assertion that `pullbackResultBuf` is the only indirect result of the pullback call.

rxwei replied:
Yes, it should be the only result. Getters are simple.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4092,6 +4180,93 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+    case StructExtractDifferentiationStrategy::Getter: {
+      // Construct pullback arguments.
+      // Allocate buffer for pullback indirect result.
+      auto *pullbackResultBuf = builder.createAllocStack(
```
I should probably change `pullbackResultBuf` to a `ValueWithCleanup` using `emitCleanup` and do cleanup before `builder.createDeallocStack(loc, pullbackResultBuf);`.

@rxwei: could you please confirm?

rxwei replied:
There's no reason to create a cleanup, because the pullback's indirect result is strictly local to the visitor. Just `emitCleanup`.

dan-zheng replied:
Thanks! Done in https://github.com/apple/swift/pull/22669/commits/9b80f4a605fcde05ad0ea8341892549982c82ba8.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4092,6 +4180,93 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+      auto *readAccess = builder.createBeginAccess(
+          loc, pullbackResultBuf, SILAccessKind::Read,
+          SILAccessEnforcement::Static, /*noNestedConflict*/ true,
+           /*fromBuiltin*/ false);
```
```suggestion
          /*fromBuiltin*/ false);
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22669/commits/9b80f4a605fcde05ad0ea8341892549982c82ba8.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4092,6 +4180,93 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+
+      // Call the pullback.
+      auto *pullbackCall = builder.createApply(
+          loc, pullback, SubstitutionMap(), {pullbackResultBuf, adjElt},
```
```suggestion
          loc, pullback, {pullbackResultBuf, adjElt},
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22669/commits/9b80f4a605fcde05ad0ea8341892549982c82ba8. Empty `SubstitutionMap` removed in other places, too.

rxwei reviewed (approved):
Excellent. Thanks for improving the transformation rules in comments.

dan-zheng committed:
Minor edits.
- Clean up local allocation for pullback indirect result before deallocation.
- Gardening.
  - Remove unnecessary empty `SubstitutionMap` argument to `SILBuilder::createApply`.

Address feedback from @rxwei.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
[AutoDiff] Fix `addToAdjointBuffer`.

Previoulsly, `addToAdjointBuffer(x, y)` called `accumulateIndirect(getAdjointBuffer(x), y)`;
`x` was assumed to be an original function buffer.

However, there are cases where `x` is not an original function buffer.
Now, `addToAdjointBuffer(x, y)` simply forwards `accumulateIndirect(x, y)`.

Add correctness assertions to aid debugging.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3653,17 +3654,19 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+    assert(destBuffer->getFunction() == &getAdjoint());
+    assert(newValueBufferAccess->getFunction() == &getAdjoint());
+    auto *destAccess = builder.createBeginAccess(
+        newValueBufferAccess.getLoc(), destBuffer, SILAccessKind::Modify,
```
Note: I think it's now wrong to always initialize `destAccess` with ` SILAccessKind::Modify`. `SILAccessKind::Init` should be used if `destBuffer` hasn't been initialized before.

(Previously, `Modify` was correct because `begin_access` was created on the output of `getAdjointBuffer(originalBuffer)`, which initializes adjoint buffers to zero and calls  `begin_access [init]`.)

I believe we can detect whether to use `Init` or `Modify` by checking if there exists a use of `destBuffer` that is a `begin_access [init]`.

@rxwei: what are your thoughts?

rxwei replied:
I don't understand why you changed `addToAdjointBuffer`'s argument's role. I'd expect it to have the old behavior.

dan-zheng replied:
Resolved in [the other comment thread](https://github.com/apple/swift/pull/22669#discussion_r257488408).

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3653,17 +3654,19 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   }
 
-  void addToAdjointBuffer(SILValue originalBuffer,
+  void addToAdjointBuffer(SILValue destBuffer,
```
I wonder if there's a better name than `addToAdjointBuffer`?
The name is not ideal because `destBuffer` may not in fact be an "adjoint buffer" (i.e. something stored in `AdjointEmitter::bufferMap`).

A more accurate name would be like "add to buffer in adjoint function". Another idea is to remove `addToAdjointBuffer` now that it simply forwards `accumulateIndirect` - but I would like to keep the function if it can be implemented with correct semantics, because it reduces caller boilerplate.

@rxwei: what are your thoughts?

rxwei replied:
Before talking about names, `addToAdjointBuffer` is supposed to take an original buffer and accumulate the new value into the adjoint buffer corresponding to it. Why did you change the behavior here?

dan-zheng replied:
Ah, that's another way to look at things. I agree with your view now.
How about this:
* Change `addToAdjointBuffer` back to the old behavior, but add assertion that `originalBuffer->getFunction() == &getAdjoint()`.
* Change current call to `addToAdjointBuffer` in `visitStructElementAddrInst` to use `accumulateIndirect`.

rxwei replied:
> * Change current call to `addToAdjointBuffer` in `visitStructElementAddrInst` to use `accumulateIndirect`.

Can you elaborate on whether `addToAdjointBuffer` doesn't work? You should call `addToAdjointBuffer` on the original operand of `visitStructElementAddrInst`.

dan-zheng replied:
> Can you elaborate on why `addToAdjointBuffer` doesn't work? You should call `addToAdjointBuffer` on the original operand of `visitStructElementAddrInst`.

Good question. The answer lies in the current transformation rule for `struct_element_addr`:
```
// Original instruction.
y = struct_element_addr x, #key
// Adjoint transformation.
adj[x]#key' = struct_element_addr adj[x], #key'
adj[x]#key' += adj[y]
// Note that `adj[x] += ...` isn't called.
```

Does this make sense to you?

dan-zheng replied:
We discussed in person and settled on this design (mentioned above):
> * Change `addToAdjointBuffer` back to the old behavior, but add assertion that `originalBuffer->getFunction() == &getAdjoint()`.
> * Change current call to `addToAdjointBuffer` in `visitStructElementAddrInst` to use `accumulateIndirect`.

This is done in https://github.com/apple/swift/pull/22669/commits/65b8c28603d2e5a31dfd62186a90a26f34639e40.

dan-zheng committed:
Revert changes to `addToAdjointBuffer`, fix `visitStructElementAddr`.

- Revert `addToAdjointBuffer` to take an `originalBuffer` operand.
  - Added correctness assertion:
    `assert(originalBuffer->getFunction() == &getOriginal())`.
- Change `visitStructElementAddr` to call `accumulateIndirect`.
- Further document transformation rules for `visitStructElementAddr`.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
